hashin.py

```diff
@@ -124,6 +149,104 @@ def get_latest_version(data):
     return data['info']['version']
 
 
+def expand_python_version(version):
+    if not re.match(r'^\d\.\d$', version):
+        return [version]
+
+    major, minor = version.split('.')
+    patterns = [
+        '{major}.{minor}',
+        'cp{major}{minor}',
+        'py{major}',
+        'py{major}.{minor}',
+        'source',
+        'py2.py3',
+    ]
+    return set(pattern.format(major=major, minor=minor) for pattern in patterns)
+
+
+# This should match the naming convention laid out in PEP 0427
+# url = 'https://pypi.python.org/packages/3.4/P/Pygments/Pygments-2.1-py3-none-any.whl'
+CLASSIFY_WHEEL_RE = re.compile('''
+    ^https://pypi.python.org/packages/[^/]+/[^/]/[^/]+/
+    (?P<package>.+)-
+    (?P<version>\d[^-]*)-
+    (?P<python_version>[^-]+)-
+    (?P<abi>[^-]+)-
+    (?P<platform>.+)
+    .(?P<format>whl)
+    (\#md5=.*)?
+    $
+''', re.VERBOSE)
+
+CLASSIFY_EGG_RE = re.compile('''
+    ^https://pypi.python.org/packages/[^/]+/[^/]/[^/]+/
+    (?P<package>.+)-
+    (?P<version>\d[^-]*)-
+    (?P<python_version>[^-]+)
+    (-(?P<platform>[^\.]+))?
+    .(?P<format>egg)
+    (\#md5=.*)?
+    $
+''', re.VERBOSE)
+
+CLASSIFY_ARCHIVE_RE = re.compile('''
+    ^https://pypi.python.org/packages/[^/]+/[^/]/[^/]+/
+    (?P<package>.+)-
+    (?P<version>\d[^-]*)
+    .(?P<format>tar.(gz|bz2)|zip)
+    (\#md5=.*)?
+    $
+''', re.VERBOSE)
```

Comment:
I took the stance of raising an error if a URL couldn't be recognized, and then kept adding regexes until I didn't get any more errors with Normandy (which I was using for my tests). I could make the classify function more forgiving and not fail on URLs it can't understand, and then this regex wouldn't be needed. Which would you prefer?